File: src/infrastructure/storage.shared/StoredAggregateEvent.cs
using System;
using core.Shared;
using Newtonsoft.Json;

namespace storage.shared
{
    public class StoredAggregateEvent
    {
        static JsonSerializerSettings _formatting = new()
        {
            TypeNameHandling = TypeNameHandling.Objects
        };

        public string Entity { get; set; }
        public Guid UserId { get; set; }
        
        public string AggregateId { get; set; }
        
        public string Key { get; set; }
        public DateTimeOffset Created { get; set; }

        public string EventJson
        {
            get => JsonConvert.SerializeObject(Event, _formatting);

            set => Event = JsonConvert.DeserializeObject<AggregateEvent>(
                EventInfraAdjustments.AdjustIfNeeded(value),
                _formatting);
        }

        public int Version { get; set; }
        public AggregateEvent Event { get; set; }
    }
}

File: src/infrastructure/storage.shared.fs/StoredAggregateEvent.fs
namespace storage.shared

open System
open core.Shared
open Newtonsoft.Json

[<CLIMutable>]
type StoredAggregateEvent = {
    Entity: string
    UserId: Guid
    AggregateId: string
    Key: string
    Created: DateTimeOffset
    Version: int
    Event: AggregateEvent
    EventJson: string
}

module StoredAggregateEvent =
    
    let private formatting = JsonSerializerSettings(TypeNameHandling = TypeNameHandling.Objects)
    
    let serializeEvent (event: AggregateEvent) : string =
        JsonConvert.SerializeObject(event, formatting)
    
    let deserializeEvent (json: string) : AggregateEvent =
        let adjustedJson = EventInfraAdjustments.AdjustIfNeeded(json)
        JsonConvert.DeserializeObject<AggregateEvent>(adjustedJson, formatting)
    
    let create entity userId aggregateId key created version event =
        {
            Entity = entity
            UserId = userId
            AggregateId = aggregateId
            Key = key
            Created = created
            Version = version
            Event = event
            EventJson = serializeEvent event
        }
